hack/update-deps.sh

```diff
@@ -20,6 +20,26 @@ set -o pipefail
 
 source $(dirname $0)/../vendor/github.com/knative/test-infra/scripts/library.sh
 
+# Remove symlinks in /vendor that are broken or lead outside the repo.
+function remove_broken_symlinks() {
+  for link in $(find ./vendor -type l); do
+    local target="$(ls -l ${link})"
+    target="${target##* -> }"
+    # Remove broken symlinks
+    if [[ ! -e ${link} ]]; then
+      unlink ${link}
+      continue
+    fi
+    # Get canonical path to target, remove if outside the repo
+    [[ ${target} == /* ]] || target="./${target}"
```

Comment:
I don't understand what's going on here, and why this is needed to find a canonical path.

I'd have this follow the links recursively. I understand we want this to work with MacOS X/BSD, so we only have a simple `readlink` and no `realpath` (which we can emulate with `$(cd $(dirname $FOO); pwd)`):

```shell
function remove_external_links() {  # Removes a chain of links pointing out of tree
  if ! [[ -e $1 ]]; then
    return 1  # This link doesn't point anywhere.
  fi
  if ! [[ -L $1 ]]; then
    return 0  # This is a real file/directory which is not a link.
  fi
  # This determines if it points outside the tree, either absolutely, or outside vendor.
  target="$(readlink "$1")"
  if [[ $target == /* ]] || [[ "$(cd $(dirname "$target"; pwd)/" == ${PWD}/vendor/*; then
    rm "$1"  # Remove the errant link, but not its target (which is outside our tree)
    return 1
  fi
  if ! remove_external_links "$target"; then
    rm "$1"  # Recursive examination found a bad target, remove this link, too.
    return 1
  fi
}

remove_external_links "${link}"
```

Reply:
FWIW, I tested this as a set of one-liners, so you may want to re-test if you decide to use it.

Note that the current script does not handle any of the following:

* chained links (`ln -s b c; ln -s a b`)
* Intermediate links that point outside the current directory (`ln -s go/src/github.com/knative/serving/vendor/foo a; ln -s ~/a go/src/github.com/knative/serving/vendor/bar`)
* Link paths which exit and re-enter the working directory. (`ln -s ../../../../../../go/src/github.com/knative/serving/vendor/foo bar`, which assumes `GOPATH=~/go` or the like. I didn't fix this one, either.)

Unfortuantely, for the third item, I think you have to process each segment of the `$target` independently and do a check as to whether it is under PWD at each step.

Reply:
a) `readlink` is not available on Mac, so this solution can't be used.
a) You don't need to check the beyond the first target. If it leads outside the repo tree, it'll be removed. If it leads to some other link inside the tree, the target link will be checked at some point.

Reply:
For b)

If you don't check beyond the first target, there may still be dangling symlinks after running N passes of this, depending on the order of directory entries returned by the OS. Consider the following:
```shell
$ cd /tmp/x
$ ls -l

b -> ../a
c -> b
d -> c

$ for x in $(find . -type l ); do [[ -L $x ]] || continue; t=$(readlink $x); [[ $t == /* ]] || t="./$t"; t2="$(cd $(dirname $x) && cd ${t%/*} && echo $PWD/${t##*/})"; echo $x $t $t2; if [[ ${t2}/ != $PWD/* ]]; then echo remove $x; continue; fi; done

./d ./c /tmp/x/c
./c ./b /tmp/x/b
./b ./../a /tmp/a
remove ./b
```

At the end of this, `d` and `c` will still be present and dangling (pointing at `b`, which was removed from the directory).

Unfortunately, I just discovered a different problem with my solution, which is that a symlink loop will cause it to recurse until stack space is exhausted.  🤷‍♂️ 

I'm okay with this solution, but wanted to point out that it might miss symlink chains where the last element in the chain points outside the tree (and that order will depend on the order that directory entries are added and returned, which is OS-dependent).

Reply:
Actually "following the symlink all the way by simple `readlink` was implemented" in a previous commit: 
https://github.com/knative/serving/pull/2842/commits/1adb644747c3dff010bf59e1c45a557f5eeaeb96

And that approach can solve the problem you mentioned, since each symlink will be solved to the canonical path of a final target (and be deleted if it is outside of the source tree). 

Let me know if you want me to revert to that approach, or do the minor improvement (as your first comment) on this approach.

Reply:
The version in 1adb644 doesn't check that the intermediate symlinks are inside the tree, allowing for e.g.

vendor/foo --> /home/evan/foo --> /home/evan/go/src/github.com/knative/serving/vendor/bar

With that said, I'm happy with a simpler solution for now and then having `dep` do the right thing.

Reply:
(And I've approved, so if people want this version, they can just `/hold cancel`.)